engine/universes.ml

```diff
@@ -14,9 +14,18 @@ open Environ
 open Univ
 open Globnames
 
-let pr_with_global_universes l =
-  try Nameops.pr_id (LMap.find l (snd (Global.global_universe_names ())))
-  with Not_found -> Level.pr l
+let reference_of_level l =
+  match Level.name l with
+  | Some (d, n as na)  ->
+     let qid =
+       try Nametab.shortest_qualid_of_universe na
+       with Not_found ->
+         let name = Id.of_string_soft (string_of_int n) in
+         Libnames.make_qualid d name
+     in Libnames.Qualid (Loc.tag @@ qid)
+  | None -> Libnames.Ident (Loc.tag @@ Id.of_string_soft (Level.to_string l))
+
```

Comment:
Why do you need a soft identifier here? Invalid identifiers should really be introduced with care and as hackish workaround against a deficient API. If indeed you can generate invalid identifiers, I'd advise to return some other type instead.

Reply:
I think it's the same `of_string_soft` that were in detyping.

Reply:
Indeed, that's a bit of a hack to make pretyping (detyping foo) = foo for funind in particular